src/channel-qvm.lisp

```diff
@@ -0,0 +1,336 @@
+   (%trial-amplitudes qvm) (make-lisp-cflonum-vector (expt 2 (number-of-qubits qvm)))
+   ;; Save a pointer to the originally provided memory
+   (slot-value qvm 'original-amplitude-pointer) (amplitudes qvm)))
+
+
+(defun requires-swapping-p (qvm)
+  "Does the noisy qvm QVM require swapping of internal pointers?"
+  (and (not (eq (amplitudes qvm) (original-amplitude-pointer qvm)))
+       #+sbcl (eq ':foreign (sb-introspect:allocation-information
+                             (original-amplitude-pointer qvm)))))
+
+
+(defmethod run :after ((qvm channel-qvm))
+  ;; Only copy if we really need to.
+  (when (requires-swapping-p qvm)
+    ;; Copy the correct amplitudes into place.
+    (copy-wavefunction (amplitudes qvm) (original-amplitude-pointer qvm))
+    ;; Get the pointer back in home position. We want to swap them,
+    ;; not overwrite, because we want the scratch memory to be intact.
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))))
+
+
+(defmethod set-readout-povm ((qvm channel-qvm) qubit povm)
+  " Set a readout povm for a qubit on this qvm"
+  (check-povm povm)
+  (setf (gethash qubit (readout-povms (noise-model qvm))) povm)
+  nil)
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
+  "If any noise rules are matched by this instruction, apply the the kraus operators for that noise rule after applying the gate for this instruction."
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+         (params (mapcar #'(lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (instr-qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+      
+         (noise-rules (noise-rules (noise-model qvm)))
+         (prepend-kraus-ops (match-rules noise-rules instr "before"))
+         (append-kraus-ops (match-rules noise-rules instr "after")))
+    (when prepend-kraus-ops (apply-all-kraus-operators qvm instr prepend-kraus-ops))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple instr-qubits) params)
+    (when append-kraus-ops (apply-all-kraus-operators qvm instr append-kraus-ops))
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod transition :around ((qvm channel-qvm) (instr quil:measurement))
+  (let ((ret-qvm (call-next-method)))
+    (apply-classical-readout-noise ret-qvm instr)
+    ret-qvm))
+
+
+(defun matches-rule (rule instr position)
+  "check if rule is matched by instruction data and the position of the match request"
+  (let* ((noise-predicate (predicate (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr) (string= position noise-position))))
+
+
+(defun match-rules (rules instr position)
+  "return the operation elements for the first matching rule"
+  (loop :for rule in rules
+        :when (matches-rule rule instr position)
+          :return (operation-elements rule)))
+
+
+(defgeneric apply-all-kraus-operators (qvm instr kraus-operators)
```

Comment:
It seems slightly odd that the generic function's lambda list here is `(qvm instr kraus-operators)`, but the method specialized on `approx-qvm` has lambda list (ignoring specializers) `(qvm instr qubits)`.

One pattern I've seen is to have the generic function accept a final `&rest args`, then each method can unpack that to whatever it's expecting: qubits for `approx-qvm` or `kraus-operators` for `channel-qvm`. Alternatively, you can get a similar effect with `&key qubits kraus-operators`, etc.

Or maybe it's normal to have methods that expect different things from their required args, and I'm just not used to seeing it. ping @stylewarning for his thoughts.

Reply:
I was able to change the lambda list for `approx-qvm` to be `(qvm instr kraus-ops)` instead of `(qvm instr qubits)` so that it matches the lambda list for the generic function and the `channel-qvm` method.